src/docfx/lib/html/HtmlWriter.cs

```diff
@@ -0,0 +1,158 @@
+            {
+                if (!reader.IsSelfClosing)
+                {
+                    _replacedTagName = reader.Name.ToString();
+                }
+                _result.Append(_replacedTag);
+                _replacedTag = null;
+            }
+            else if (_replacedTagName != null)
+            {
+                if (reader.Type == HtmlTokenType.EndTag && reader.NameIs(_replacedTagName))
+                {
+                    _replacedTagName = null;
+                }
+            }
+            else if (_replacedToken != null)
+            {
+                _result.Append(_replacedToken);
+            }
+            else if (_newAttributes.Length == 0 && _replacedAttributes.Length == 0 && _replacedAttributeValues.Length == 0)
+            {
+                _result.Append(reader.Token);
+            }
+            else
+            {
+                _result.Append('<').Append(reader.Name);
+
+                var pos = reader.TokenRange.start + reader.Name.Length + 1;
+
+                foreach (ref readonly var attribute in reader.Attributes)
+                {
+                    var replaced = false;
+
+                    foreach (var (start, value) in _replacedAttributes.Span)
+                    {
+                        var tokenRange = attribute.TokenRange;
+                        if (tokenRange.start == start)
+                        {
+                            _result.Append(reader.Html, pos, start - pos);
+                            _result.Append(value);
+                            pos = tokenRange.start + tokenRange.length;
+                            replaced = true;
+                            break;
+                        }
+                    }
+
+                    if (!replaced)
+                    {
+                        foreach (var (start, value) in _replacedAttributeValues.Span)
+                        {
+                            var valueRange = attribute.ValueRange;
+                            if (valueRange.start == start)
+                            {
+                                _result.Append(reader.Html, pos, start - pos);
+                                _result.Append(value);
+                                pos = valueRange.start + valueRange.length;
+                                break;
+                            }
+                        }
+                    }
+                }
+
+                foreach (var attribute in _newAttributes.Span)
+                {
+                    _result.Append(' ').Append(attribute).Append(' ');
+                }
+
+                var end = reader.TokenRange.start + reader.TokenRange.length;
+                _result.Append(reader.Html, pos, end - pos);
+            }
+
+            // Reset
+            _replacedToken = null;
+            _newAttributes.Clear();
+            _replacedAttributes.Clear();
+            _replacedAttributeValues.Clear();
+        }
+
+        public void ReplaceTag(string tag)
```

Comment:
This is the only operation method in use, why are we having the others in this PR?

Reply:
For completeness. The others will be used later.